src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -181,7 +196,7 @@ suggestRangeFinderPrecision maxPages =
 --
 -- TODO: optimisation ideas: we should be able to get it all unboxed,
 -- non-allocating and without unnecessary bounds checking.
-search :: SerialisedKey -> CompactIndex -> Maybe Int
+search :: SerialisedKey -> CompactIndex -> SearchResult
```

Comment:
`gallopingSearchLeftPBounds` can return `ub` in case the predicate returns `False` for all elements. However, since we're subtracting `1`, it's actually the interval `[lb-1, ub)`. In any case, I'm going to follow your advice to factor out this search code and be more precise about what the function does